design/design-doc-agent.md

```diff
@@ -0,0 +1,627 @@
+  name: agent-requirement
+  namespace: foo
+rules:
+- apiGroups: [""]
+  resources: ["secrets"]
+  verbs: ["*"]
+- apiGroups: ["database.acmeco.org"]
+  resources: ["*"]
+  verbs: ["*"]
+- apiGroups: ["network.acmeco.org"]
+  resources: ["*"]
+  verbs: ["*"]
+---
+apiVersion: rbac.authorization.k8s.io/v1
+kind: RoleBinding
+metadata:
+  name: agent-requirement
+  namespace: foo
+subjects:
+- kind: ServiceAccount
+  name: agent1
+  apiGroup: rbac.authorization.k8s.io
+roleRef:
+  kind: Role
+  name: agent-requirement
+  apiGroup: rbac.authorization.k8s.io
+```
+
+The YAML above includes what's necessary to sync a specific namespace. The YAML
+below is for cluster-scoped resources that should be read by the agent and it's
+generic to be used by all agents except the `subjects` list of
+`ClusterRoleBinding`:
+
+```yaml
+# To be able to read the cluster-scoped resources.
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRole
+metadata:
+  name: read-for-agents
+rules:
+- apiGroups: ["apiextensions.kubernetes.io"]
+  resources: ["customresourcedefinitions"]
+  verbs: ["get", "watch", "list"]
+- apiGroups: ["apiextensions.crossplane.io"]
+  resources:
+  - infrastructuredefinitions
+  - infrastructurepublications
+  - compositions
+  verbs: ["get", "watch", "list"]
+---
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRoleBinding
+metadata:
+  name: read-for-agents
+subjects:
+- kind: ServiceAccount
+  name: agent1
+  apiGroup: rbac.authorization.k8s.io
+roleRef:
+  kind: ClusterRole
+  name: read-for-agents
+  apiGroup: rbac.authorization.k8s.io
+```
+
+At this point, we have a `ServiceAccount` with all necessary permissions in our
+central cluster. You can think of it like IAM user in the public cloud
+providers; we have created it and allowed it to access a certain set of APIs.
+Later on, its key will be used by the agent; just like provider-aws using the
+key of an IAM User.
+
+1. User provisions a network stack and a Kubernetes cluster in that private
+   network through Crossplane (or via other methods, even kind cluster would
+   work). This cluster will be used as remote cluster. We'll run the rest of the
+   steps in that remote cluster.
+1. The `Secret` that contains the kubeconfig of the `ServiceAccount` we created
+   in the central cluster is replicated in the remote cluster with name
+   `agent1-creds`.
+1. The agent is installed into the remote cluster. The Helm package will have
+   all the necessary RBAC resources but user will need to enter the API groups
+   of the published types so that it can reconcile them in the remote cluster.
+   `helm install crossplane/agent --set apiGroups=database.acmeco.org,network.acmeco.org --set credentials-secret-name=agent1-creds`
+1. The `Namespace`s you'd like to sync are created with the necessary annotation
+   like the following:
+
+  ```yaml
+  apiVersion: v1
+  kind: Namespace
+  metadata:
+    name: foo
+    annotations:
+      "agent.crossplane.io/target-namespace": bar
+  ```
```

Comment:
Again, is this on the remote or the crossplane cluster? My guess is that it's on the remote but why is the ns name "foo" which is the same as the on the central cluster above.

Reply:
Yeah, it's on the remote. I corrected the names of the namespaces, thanks!